Win32.xs

```diff
@@ -1682,6 +1685,307 @@ XS(w32_IsDeveloperModeEnabled)
+        bResults = WinHttpQueryHeaders(hRequest,
+                                       WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER,
+                                       WINHTTP_HEADER_NAME_BY_INDEX,
+                                       &dwHttpStatusCode,
+                                       &dwQuerySize,
+                                       WINHTTP_NO_HEADER_INDEX);
+    }
+
+    /* Retrieve HTTP status text. Note this may be a success message. */
+    if (bResults) {
+        dwQuerySize = ONE_K_BUFSIZE * 2 - 2;
+        ZeroMemory(&msgbuf, ONE_K_BUFSIZE * 2);
+        bResults = WinHttpQueryHeaders(hRequest,
+                                       WINHTTP_QUERY_STATUS_TEXT,
+                                       WINHTTP_HEADER_NAME_BY_INDEX,
+                                       msgbuf,
+                                       &dwQuerySize,
+                                       WINHTTP_NO_HEADER_INDEX);
+    }
+
+    /* There is no point in successfully downloading an error page from
+     * the server, so consider HTTP errors to be failures.
+     */
+    if (bResults) {
+        if (dwHttpStatusCode < 200 || dwHttpStatusCode > 299) {
+            bResults = FALSE;
+            bHttpError = TRUE;
+        }
+    }
+
+    /* Create output file for download. */
+    if (bResults) {
+        hOut = CreateFileW(file,
+                           GENERIC_WRITE,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE,
+                           NULL,
+                           CREATE_ALWAYS,
+                           FILE_ATTRIBUTE_NORMAL,
+                           NULL);
+
+        if (hOut == INVALID_HANDLE_VALUE)
+            bFileError = TRUE;
+    }
+
+    if (!bFileError && bResults) {
+        DWORD dwDownloaded = 0;
+        DWORD dwBytesWritten = 0;
+        DWORD dwSize = 65536;
+        char *pszOutBuffer;
+
+        New(0, pszOutBuffer, dwSize, char);
+
+        /* Keep checking for data until there is nothing left. */
+        while (1) {
+            if (!WinHttpReadData(hRequest,
+                                 (LPVOID)pszOutBuffer,
+                                 dwSize,
+                                 &dwDownloaded)) {
+                bAborted = TRUE;
+                break;
+            }
+            if (!dwDownloaded)
+                break;
+
+            /* Write what we just read to the output file */
+            if (!WriteFile(hOut,
+                           pszOutBuffer,
+                           dwDownloaded,
+                           &dwBytesWritten,
+                           NULL)) {
+                bAborted = TRUE;
+                bFileError = TRUE;
+                break;
+            }
+
+        }
+
+        Safefree(pszOutBuffer);
+    }
+    else {
+        bAborted = TRUE;
+    }
+
+    /* Clean-up may lose this. */
+    if (bAborted)
+        error = GetLastError();
+
+    /* Close any open handles. */
+    if (hOut) CloseHandle(hOut);
```

Comment:
If `WinHttpReadData` or the `WriteFile` calls failed above (so `bAborted` is true here), then the file still needs to be deleted because it is empty of incomplete.

Reply:
> If `WinHttpReadData` or the `WriteFile` calls failed above (so `bAborted` is true here), then the file still needs to be deleted because it is empty or incomplete.

I've added a delete.  I can't think of a way to test it because you'd have to have a disk fill up after successfully opening the file or a network connection drop after starting the transfer or something like that.